IA-1861 Overhaul of the Completeness Stats page.

olethanh opened the pull request:
IA-1861
Redid the whole backend code so this is computed in database which make the whole thing much faster. Adapted the UX. added requested filter.




## Changes
- Simplified the org unit filter logic. IA-1919, UA-1755
- Columns for each Form (instead of rows)
- New filter on period. IA-1942
- Filter orgunit belonging on a group (in the count) IA-1918
- New filter on planning. Note that this doesn't filter the orgunit count, just the submission. IA-1942
- Ordering of column now working. IA-1792
- Respect orgunit limitation for user IA-1901
- The seed command was also changed to properly set orgunit type on created orgunit. IA-2045
- Add CSV export
- Reworker period picker budget a bit

## Self proofreading checklist

- [x] Did I use eslint and black formatters
- [x] Is my code clear enough and well documented
- [x] Are my typescript files well typed
- [x] New translations have been added or updated if new string have been introduced in the frontend. TODO
- [NA] My migrations file are included
- [no] Are there enough tests
- [ ] Documentation has been included (for new feature). TODO

## How to test
Have data in your Iaso DB, use both way:
1.
Try using it with your existing database. If you don't have submission but not any data in the table it's probably that the on the Form the orgunit type settings is not correctly set.

2. 
Use the seed command to fill your Database with OrgUnit from DHIS2 and Submission.

Go with the page, play with sorting, filter. Try creating a group and putting orgunit in it


## Print screen / video

https://user-images.githubusercontent.com/82500/229712467-3809d6ba-1c8e-407f-8759-4426b4bdf706.mp4




## Notes
A new python dependency has been added, so you will need to rebuild you django Django image. I'm also using the latest bluesquare-component version from Son that added the rowProps, already merged in main iaso but if you didn't update you image yet you will need to do it.

If you have suggestions on how to improve the columns cell I'm listening. I should add more link in them.

olethanh commented:
Thanks for adding the translation I forgot last time.

olethanh committed:
WIP

olethanh committed:
in the select

olethanh committed:
wip

olethanh committed:
Adapt ui, this start to look like something

olethanh committed:
Show stat for selectd org unit

olethanh committed:
cleaning

olethanh committed:
Form on parent: filter with qs. Accept ou in new state

olethanh committed:
OrgUnit Type filter now group by properly

change terminology to make clear what it does
Add modification so it filter automatically which org unit type are shown

olethanh committed:
Handle invaliding orgunit type

olethanh committed:
fix warning typing

olethanh committed:
WIP planning filteR

olethanh committed:
fix counting submission on root

olethanh committed:
use orgunitlevels as type in import

olethanh committed:
fix order by

but still no order by stats

olethanh committed:
ComletenessStats: Period filter use type corresponding to Form

olethanh committed:
typo

olethanh committed:
ComletenessStats: Add planning filter

olethanh committed:
Add csv extract api

olethanh committed:
Add a link to csv download

olethanh committed:
WIP rewrite the whole thing as cte so we can filter

olethanh committed:
Reimplement itself__

olethanh committed:
Reimplement filtering for period and planning_id

olethanh committed:
Use a serializer for validating params

olethanh committed:
fix bug whem empty result

olethanh committed:
fix filter button

olethanh committed:
Merge remote-tracking branch 'origin/main' into comp

olethanh committed:
Display requested parent in it's own top line

olethanh committed:
Hide requested orgunit stats box

olethanh committed:
On the row for requested parent make the link go to the top level

olethanh committed:
clean code, remove now unused request_parent_forms_stats

olethanh committed:
Show the percentage on top

olethanh committed:
code mod

olethanh committed:
IA-1792 Completeness add sorting on form stats

olethanh committed:
Clean style and design of table

olethanh committed:
Python code cleaning

olethanh committed:
fix mypy error

olethanh committed:
code cleaning

olethanh committed:
CompStat api: Allow filtering on group

olethanh committed:
CompStats: Frontend, show filter for group

olethanh committed:
Fix typing

olethanh committed:
Allow filtering per parent name

olethanh committed:
CompStat use a keepPreviousData

this way this feel smooth when using the button
othewhise when navigating it display an empty table then a filled one

olethanh committed:
Fix comment and title

olethanh committed:
fix bug in group by type

olethanh committed:
better typing

olethanh committed:
remove old code

olethanh committed:
Adapt test to the new api

beygorghor commented on hat/assets/js/apps/Iaso/domains/completenessStats/CompletenessStatsFilters.tsx:
```diff
@@ -12,6 +18,19 @@ import { useGetOrgUnit } from '../orgUnits/components/TreeView/requests';
+    PERIOD_TYPE_QUARTER,
+    PERIOD_TYPE_YEAR,
+} from '../periods/constants';
+import _ from 'lodash';
```
Please import, only what you need form lodash not the whole library.

beygorghor commented on hat/assets/js/apps/Iaso/domains/completenessStats/hooks/api/useGetOrgUnitTypesOptions.ts:
```diff
@@ -1,26 +1,43 @@
 import { getRequest } from '../../../../libs/Api';
 import { useSnackQuery } from '../../../../libs/apiHooks';
+import { memoize } from 'lodash';
+import { UseQueryResult } from 'react-query';
```
This should be at the top of the file

beygorghor commented on hat/assets/js/apps/Iaso/domains/completenessStats/hooks/api/useGetOrgUnitTypesOptions.ts:
```diff
@@ -1,26 +1,43 @@
-    const queryString = new URLSearchParams(params);
+    const filteredParams = Object.entries(params).filter(
+        // eslint-disable-next-line no-unused-vars
+        ([_key, value]) => value !== undefined,
```
if you use only `_`, no need to disable eslint

olethanh replied:
eslint still make me an error if I don't disable it, am I missing a config?

beygorghor commented on hat/assets/js/apps/Iaso/domains/completenessStats/hooks/api/useGetOrgUnitTypesOptions.ts:
```diff
@@ -1,26 +1,43 @@
+};
+
+export const useGetOrgUnitTypesOptions: (
+    number?,
```
This is not used

beygorghor commented on hat/assets/js/apps/Iaso/domains/completenessStats/hooks/useCompletenessStatsColumns.tsx:
```diff
@@ -1,32 +1,64 @@
-export const useCompletenessStatsColumns = (params: any) => {
+// From https://v4.mui.com/components/progress/ to clean
+const LinearProgressWithLabel = props => (
+    <Box display="flex" alignItems="center" flexDirection={'column'}>
```
`flexDirection={'column'}` => `flexDirection="column"`

beygorghor commented on hat/assets/js/apps/Iaso/domains/completenessStats/hooks/useCompletenessStatsColumns.tsx:
```diff
@@ -1,32 +1,64 @@
+        <Box minWidth={35}>
+            <Typography variant="body2" color="textSecondary">{`${Math.round(
+                props.value,
+            )}%`}</Typography>
```
indent issue

beygorghor commented on hat/assets/js/apps/Iaso/domains/completenessStats/hooks/useCompletenessStatsColumns.tsx:
```diff
@@ -1,32 +1,64 @@
+export type FormStatRow = {
+    value: FormStat;
+};
+export const useCompletenessStatsColumns = (params: any, completenessStats) => {
```
`completenessStats` is not typed

beygorghor commented on hat/assets/js/apps/Iaso/domains/completenessStats/hooks/useCompletenessStatsColumns.tsx:
```diff
@@ -56,107 +88,153 @@ export const useCompletenessStatsColumns = (params: any) => {
+                                        </>
+                                    ) : (
+                                        <div
+                                            title={
```
`title="..."`

beygorghor commented on hat/assets/js/apps/Iaso/domains/completenessStats/hooks/useCompletenessStatsColumns.tsx:
```diff
@@ -56,107 +88,153 @@ export const useCompletenessStatsColumns = (params: any) => {
+                                        <>
+                                            {value.itself_has_instances ? (
+                                                <span
+                                                    title={`Total #submissions: ${value.itself_instances_count}`}
```
`title="..."`

beygorghor commented on hat/assets/js/apps/Iaso/domains/completenessStats/hooks/useCompletenessStatsColumns.tsx:
```diff
@@ -56,107 +88,153 @@ export const useCompletenessStatsColumns = (params: any) => {
+                            //     accessor: `form_stats[${form.slug}].total_instances`,
+                            //     Cell: ({ value }) => <>{value}</>,
+                            //     sortable: false,
+                            // },
```
🗑️

beygorghor commented on hat/assets/js/apps/Iaso/domains/completenessStats/hooks/useCompletenessStatsColumns.tsx:
```diff
@@ -171,11 +249,28 @@ export const useCompletenessStatsColumns = (params: any) => {
-        [dispatch, formatMessage, redirectionParams],
-    );
+        });
+        console.log(columns);
```
console.log

beygorghor commented on hat/assets/js/apps/Iaso/domains/completenessStats/index.tsx:
```diff
@@ -30,13 +32,29 @@ type Props = {
+    if (original?.is_root) {
+        return {
+            style: {
+                backgroundColor: '#b1e5e5',
```
can we use theme color or at least a material ui color ?

olethanh replied:
which do you suggest?

beygorghor commented on hat/assets/js/apps/Iaso/domains/completenessStats/index.tsx:
```diff
@@ -48,12 +66,21 @@ export const CompletessStats: FunctionComponent<Props> = ({ params }) => {
+                    item
+                    style={{ paddingTop: '5px', paddingBottom: '5px' }}
+                >
+                    <Grid item container justifyContent={'flex-end'}>
```
`justifyContent="..."`

beygorghor commented on hat/assets/js/apps/Iaso/domains/completenessStats/types.ts:
```diff
@@ -12,4 +12,11 @@ export type CompletenessStats = {
+        name: string;
+        id: number;
+        slug: string;
+    }[];
```
we already have a Form type somewhere no ?

olethanh replied:
This one is a special one constructed by the api. because for example `slug` exist only in this context.

beygorghor commented on hat/assets/js/apps/Iaso/types/utils.ts:
```diff
@@ -6,6 +6,11 @@ export type DropdownOptions<T> = {
     value: T;
     original?: Record<any, any>;
 };
+export type DropdownOptionsWithOriginal<T> = {
+    label: string;
+    value: T;
+    original: Record<any, any>;
+};
```
not sure why you need this ? if you really need original has not optional, maybe extend `DropdownOptions` instead

olethanh replied:
I got an type error when using DropdownOptions because original was not garanteed  to be there

beygorghor commented on hat/assets/js/apps/Iaso/domains/completenessStats/hooks/api/useGetCompletnessStats.ts:
```diff
@@ -4,24 +4,35 @@ import { useSnackQuery } from '../../../../libs/apiHooks';
-const getCompletenessStats = async (
-    params: CompletenessGETParams,
-): Promise<CompletenessApiResponse> => {
+export const buildQueryString = (
```
return type is missing

beygorghor commented on hat/assets/js/apps/Iaso/domains/completenessStats/CompletenessStatsFilters.tsx:
```diff
@@ -12,6 +18,19 @@ import { useGetOrgUnit } from '../orgUnits/components/TreeView/requests';
+    PERIOD_TYPE_MONTH,
+    PERIOD_TYPE_QUARTER,
+    PERIOD_TYPE_YEAR,
+} from '../periods/constants';
```
useless

beygorghor commented on hat/assets/js/apps/Iaso/domains/completenessStats/CompletenessStatsFilters.tsx:
```diff
@@ -12,6 +18,19 @@ import { useGetOrgUnit } from '../orgUnits/components/TreeView/requests';
+} from '../periods/constants';
+import _ from 'lodash';
+import {
+    useGetPlannings,
```
useless

beygorghor commented on hat/assets/js/apps/Iaso/domains/completenessStats/CompletenessStatsFilters.tsx:
```diff
@@ -34,18 +54,65 @@ export const CompletenessStatsFilters: FunctionComponent<Props> = ({
+            const out: string = filters.orgUnitTypeIds as string;
+            const selectedOrgUnitIDs = out
+                .split(',')
+                .map(x => parseInt(x))
```
please use radix

beygorghor commented on hat/assets/js/apps/Iaso/domains/completenessStats/CompletenessStatsFilters.tsx:
```diff
@@ -34,18 +54,65 @@ export const CompletenessStatsFilters: FunctionComponent<Props> = ({
+            const formIdStr = filters.formId as string;
+            const selectedFormIds = formIdStr
+                .split(',')
+                .map(x => parseInt(x))
```
radix

beygorghor commented on hat/assets/js/apps/Iaso/domains/completenessStats/CompletenessStatsFilters.tsx:
```diff
@@ -34,18 +54,65 @@ export const CompletenessStatsFilters: FunctionComponent<Props> = ({
+                .map(x => parseInt(x))
+                .sort();
+            const selectedForms = forms.filter(
+                f => selectedFormIds.indexOf(f.value) != -1,
```
!==

beygorghor commented on hat/assets/js/apps/Iaso/domains/completenessStats/CompletenessStatsFilters.tsx:
```diff
@@ -34,18 +54,65 @@ export const CompletenessStatsFilters: FunctionComponent<Props> = ({
+                .split(',')
+                .map(x => parseInt(x))
+                .sort();
+            const availableIds = orgUnitTypes.map(out => out.value);
```
be careful `out` is already declared

beygorghor commented on hat/assets/js/apps/Iaso/domains/completenessStats/CompletenessStatsFilters.tsx:
```diff
@@ -34,18 +54,65 @@ export const CompletenessStatsFilters: FunctionComponent<Props> = ({
+
+            const periods = selectedForms
+                .map(f => f.original.period_type)
+                .filter(period_type => Boolean(period_type));
```
`periodType`

beygorghor commented on hat/assets/js/apps/Iaso/domains/completenessStats/CompletenessStatsFilters.tsx:
```diff
@@ -34,18 +54,65 @@ export const CompletenessStatsFilters: FunctionComponent<Props> = ({
+            return uniqPeriods ? uniqPeriods[0] : null;
+        }
+        return null;
+    }, [filters]);
```
`forms` is missing ? if this is the wanted effect, please disable eslint

beygorghor commented on hat/assets/js/apps/Iaso/domains/completenessStats/CompletenessStatsFilters.tsx:
```diff
@@ -34,18 +54,65 @@ export const CompletenessStatsFilters: FunctionComponent<Props> = ({
+        return null;
+    }, [filters]);
 
     const handleOrgUnitChange = useCallback(
```
this is not used anymore

beygorghor commented on hat/assets/js/apps/Iaso/domains/completenessStats/CompletenessStatsFilters.tsx:
```diff
@@ -73,14 +140,33 @@ export const CompletenessStatsFilters: FunctionComponent<Props> = ({
+                    {/*        onConfirm={handleOrgUnitChange}*/}
+                    {/*        initialSelection={initialOrgUnit}*/}
+                    {/*    />*/}
+                    {/*</Box>*/}
```
eslint

beygorghor commented on hat/assets/js/apps/Iaso/domains/completenessStats/CompletenessStatsFilters.tsx:
```diff
@@ -73,14 +140,33 @@ export const CompletenessStatsFilters: FunctionComponent<Props> = ({
+                    {/*</Box>*/}
+                </Grid>
+                <Grid item xs={12} md={3}>
+                    {/*FIXME Connect to the rest*/}
```
eslint

beygorghor commented on hat/assets/js/apps/Iaso/domains/completenessStats/CompletenessStatsFilters.tsx:
```diff
@@ -73,14 +140,33 @@ export const CompletenessStatsFilters: FunctionComponent<Props> = ({
+                        // activePeriodString={formState.startPeriod.value}
+                        // periodType={formState.periodType.value}
+                        periodType={periodType}
+                        title={'Period'}
```
`title="..."`

beygorghor commented on hat/assets/js/apps/Iaso/domains/completenessStats/CompletenessStatsFilters.tsx:
```diff
@@ -73,14 +140,33 @@ export const CompletenessStatsFilters: FunctionComponent<Props> = ({
+                        periodType={periodType}
+                        title={'Period'}
+                        // title={formatMessage(MESSAGES.startPeriod)}
+                        keyName={'period'}
```
...

beygorghor commented on hat/assets/js/apps/Iaso/domains/completenessStats/CompletenessStatsFilters.tsx:
```diff
@@ -91,29 +177,57 @@ export const CompletenessStatsFilters: FunctionComponent<Props> = ({
-                    </Box>
+                    </Grid>
+                    <DisplayIfUserHasPerm
+                        permission={'iaso.permissions.planning'}
```
`permissions="..."`

beygorghor reviewed (changes requested):
Eslint and typescript issues

olethanh committed:
fix code, review comment, eslint

olethanh committed:
More code cleaning, remove old code eslint

olethanh committed:
show periodpicker even if no periodType yet

fix look of PeriodPicker:
  border on hover
  style of label

olethanh committed:
Add structure to allow translations

olethanh committed:
Add translations

olethanh committed:
Adapt the python test for the modificaiton in the dhis2 import command

beygorghor committed:
sort messages by alphanum

olethanh committed:
fix crash

beygorghor committed:
small fixes

olethanh committed:
remove comments

olethanh committed:
fix form with no periodType

olethanh committed:
invert group and period picker

beygorghor committed:
filter rearrange

olethanh committed:
Try to improve perfs

limit number of form per default to not have too much column
pre filter orgunits before calculating

olethanh committed:
Fix test error in CI

this was caused by float imprecison
use a method that check if they are close enough

this is needed because unittest has a self.assertAlmostEqual
but it is not recursive

olethanh committed:
idem

olethanh committed:
idem idem

olethanh committed:
fix ci

olethanh committed:
fix mypi

olethanh committed:
more typing

olethanh committed:
Derivate is_root color from style

olethanh committed:
fix spelling

beygorghor commented on hat/assets/js/apps/Iaso/constants/routes.js:
```diff
@@ -539,7 +539,7 @@ export const completenessPath = {
     baseUrl: baseUrls.completenessStats,
     permissions: ['iaso_completeness_stats'],
-    component: props => <CompletessStats {...props} />,
+    component: props => <CompletenessStats {...props} />,
```
🤣

beygorghor reviewed (commented):
GG

olethanh committed:
Filtering per mulitple type
